server/jetstream_test.go

```diff
@@ -10392,6 +10392,362 @@ func TestJetStreamPullConsumerMaxAckPendingRedeliveries(t *testing.T) {
+		}
+
+		// Make a few requests to start getting 408 Request Timeout errors
+		// due to reaching the limit.
+		for i := 0; i < 4; i++ {
+			msgs, err = sub.Fetch(2, nats.MaxWait(200*time.Millisecond))
+			if err == nil {
+				t.Error("Unexpected success")
+			}
+			if len(msgs) != 0 {
+				t.Error("Expected no messages")
+			}
+
+			// There should be a maximum number of waiting requests now.
+			info, err = sub.ConsumerInfo()
+			if err != nil {
+				t.Fatal(err)
+			}
+			pending += 1
+			if info.NumWaiting != pending {
+				t.Errorf("Expected %v pending requests, got: %v", i+2, info.NumWaiting)
+			}
+		}
+
+		// There should be a maximum number of waiting requests now.
+		info, err = sub.ConsumerInfo()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if info.NumWaiting != 5 {
+			t.Errorf("Expected 5 pending requests, got: %v", info.NumWaiting)
+		}
+
+		// Making an extra request will expire some of the old requests.
+		msgs, err = sub.Fetch(2, nats.MaxWait(200*time.Millisecond))
+		if err == nil {
+			t.Error("Unexpected success")
+		}
+		if len(msgs) != 0 {
+			t.Error("Expected no messages")
+		}
+		info, err = sub.ConsumerInfo()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if info.NumWaiting != 1 {
+			t.Errorf("Expected 1 pending requests, got: %v", info.NumWaiting)
+		}
+
+		// Send another message with a delay...
+		time.AfterFunc(200*time.Millisecond, func() {
+			js.Publish(subject, []byte("bar"))
+		})
+		msgs, err = sub.Fetch(1, nats.MaxWait(500*time.Millisecond))
+		if err != nil {
+			t.Error(err)
+		}
+		msg = msgs[0]
+		expected = "bar"
+		got = string(msg.Data)
+		if got != expected {
+			t.Errorf("Expected: %v, got: %v", expected, got)
+		}
+		if len(msgs) != 1 {
+			t.Fatalf("Expected one message to be delivered, got: %v", len(msgs))
+		}
+
+		// There should be no waiting pull requests since they got
+		// expired after fetch succeeded.
+		info, err = sub.ConsumerInfo()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if info.NumWaiting != 0 {
+			t.Errorf("Expected no pending requests, got: %v", info.NumWaiting)
+		}
+	})
+
+	t.Run("blocking fetch", func(t *testing.T) {
+		// Create requests that take a longer time and will exhaust
+		// the number of waiting requests so that the rest will be blocked.
+		var (
+			max         = 5
+			msgCh       = make(chan *nats.Msg, max)
+			errCh       = make(chan error, max)
+			expectedMap = make(map[string]bool)
+		)
+
+		ctx, cancelFetch := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancelFetch()
+
+		for i := 0; i < max; i++ {
+			expectedMap[fmt.Sprintf("quux:%d", i)] = false
+			go func() {
+				// Can only have at most 5 inflight fetch requests so
+				// so any fetch will timeout until these are canceled.
+				// These may receive some messages but they are not going
+				// to return until the context cancels the request forcefully later.
+				msgs, err := sub.Fetch(1000, nats.Context(ctx))
```

Comment:
These requests will block for 30 seconds any activity on the consumer, so new Fetch requests will not work and return `nats: timeout` until either 30 seconds pass or 1000 messages are received.

Reply:
Might be worth discussing a bit, maybe we can all jump on a short call tomorrow.